state/pruner.go

```diff
@@ -0,0 +1,341 @@
+}
+
+// GetABCIResRetainHeight is a convenience method for accessing the
+// GetABCIResRetainHeight method of the underlying state store.
+func (p *Pruner) GetABCIResRetainHeight() (int64, error) {
+	return p.stateStore.GetABCIResRetainHeight()
+}
+
+func (p *Pruner) pruneABCIResRoutine() {
+	p.logger.Info("Started pruning ABCI results", "interval", p.interval.String())
+	lastABCIResRetainHeight := int64(0)
+	for {
+		select {
+		case <-p.Quit():
+			return
+		default:
+			newABCIResRetainHeight := p.pruneABCIResToRetainHeight(lastABCIResRetainHeight)
+			p.observer.PrunerPrunedABCIRes(&ABCIResponsesPrunedInfo{
+				FromHeight: lastABCIResRetainHeight,
+				ToHeight:   newABCIResRetainHeight - 1,
+			})
+			lastABCIResRetainHeight = newABCIResRetainHeight
+			time.Sleep(p.interval)
+		}
+	}
+}
+func (p *Pruner) pruneBlocksRoutine() {
+	p.logger.Info("Pruner started", "interval", p.interval.String())
+	lastRetainHeight := int64(0)
+	for {
+		select {
+		case <-p.Quit():
+			return
+		default:
+			newRetainHeight := p.pruneBlocksToRetainHeight(lastRetainHeight)
+			p.observer.PrunerPrunedBlocks(&BlocksPrunedInfo{
+				FromHeight: lastRetainHeight,
+				ToHeight:   newRetainHeight - 1,
+			})
+			lastRetainHeight = newRetainHeight
+			time.Sleep(p.interval)
+		}
+	}
+}
+
+func (p *Pruner) pruneBlocksToRetainHeight(lastRetainHeight int64) int64 {
+	targetRetainHeight := p.findMinRetainHeight()
+	if targetRetainHeight == lastRetainHeight {
+		return lastRetainHeight
+	}
+	pruned, evRetainHeight, err := p.pruneBlocks(targetRetainHeight)
+	// The new retain height is the current lowest point of the block store
+	// indicated by Base()
+	newRetainHeight := p.bs.Base()
+	if err != nil {
+		p.logger.Error("Failed to prune blocks", "err", err, "targetRetainHeight", targetRetainHeight, "newRetainHeight", newRetainHeight)
+	} else if pruned > 0 {
+		p.logger.Info("Pruned blocks", "count", pruned, "evidenceRetainHeight", evRetainHeight, "newRetainHeight", newRetainHeight)
+	}
+	return newRetainHeight
+}
+
+func (p *Pruner) pruneABCIResToRetainHeight(lastRetainHeight int64) int64 {
+	targetRetainHeight, err := p.stateStore.GetABCIResRetainHeight()
+	if err != nil {
+		// ABCI response retain height has not yet been set - do not log any
+		// errors at this time.
+		if errors.Is(err, ErrKeyNotFound) {
+			return 0
+		}
+		p.logger.Error("Failed to get ABCI result retain height", "err", err)
+		return lastRetainHeight
+	}
+
+	if lastRetainHeight == targetRetainHeight {
+		return lastRetainHeight
+	}
+
+	// newRetainHeight is the height just after that which we have successfully
+	// pruned. In case of an error it will be 0, but then it will also be
+	// ignored.
+	numPruned, newRetainHeight, err := p.stateStore.PruneABCIResponses(targetRetainHeight)
+	if err != nil {
+		p.logger.Error("Failed to prune ABCI responses", "err", err, "targetRetainHeight", targetRetainHeight)
+		return lastRetainHeight
+	}
+	p.logger.Info("Pruned ABCI responses", "height", numPruned)
```

Comment:
```suggestion
	if numPruned > 0 {
		p.logger.Info("Pruned ABCI responses", "heights", numPruned, "newRetainHeight", newRetainHeight)
	}
```

So as to not spam operators.